src/AdoPat/PatManager.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AdoPat
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.Services.DelegatedAuthorization;
+
+    /// <summary>
+    /// Manage the interaction between the PAT cache and Azure DevOps.
+    /// </summary>
+    public class PatManager
+    {
+        private const int ValidToExtensionDays = 7;
+        private const int NearingExpirationDays = 2;
+
+        private IPatCache cache;
+        private IPatClient client;
+        private Func<DateTime> now;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PatManager"/> class.
+        /// </summary>
+        /// <param name="cache">Any class that implements the <see cref="IPatCache"/> interface.</param>
+        /// <param name="client">Any class that implements the <see cref="IPatClient"/> interface.</param>
+        /// <param name="now">A function for computing the current moment. Defaults to null, which
+        /// uses <see cref="DateTime.UtcNow"/>. Overriding this should only be necessary in testing.</param>
+        public PatManager(IPatCache cache, IPatClient client, Func<DateTime> now = null)
+        {
+            this.cache = cache;
+            this.client = client;
+            this.now = now ?? new Func<DateTime>(() => DateTime.UtcNow);
+        }
+
+        /// <summary>
+        /// Fetches a PAT matching the given options from the cache, creating or regenerating a new PAT as necessary.
+        /// </summary>
+        /// <param name="options">Options used to match the PAT in the cache or when creating/regenerating a new one.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>An Azure DevOps Personal Access Token.</returns>
+        public async Task<PatToken> GetPatAsync(
+            PatOptions options,
+            CancellationToken cancellationToken = default)
+        {
+            bool writeBack = false;
+            var pat = this.cache.GetPat(options.ToString());
+
+            if (pat == null || await this.Inactive(pat, cancellationToken).ConfigureAwait(false))

```

Comment:
In general this method feels a little awkward because we're first seeing if we have a cached pat. Then we check if it's null, or inactive, and if so Create a new one. Then we immeditaely check if that potentially new PAT is close to expiring. I feel like the logic should split or could be more streramlined here, such that we don't check if a PAT we know we just made is about to expire.

Reply:
I did it this way because I wanted to minimize the number of times we called `PutPat`, but I can search for a less "awkward" implementation and see what I can come up with.

Reply:
The simplest thing I can think of is to drop the concept of `writeBack` for an `else if` and write to the cache in both arms of the conditional.
```csharp
var pat = this.cache.GetPat(options.ToString());

if (pat == null || await this.Inactive(pat, cancellationToken).ConfigureAwait(false))
{
    var patTokencreateRequest = new PatTokenCreateRequest
    {
        DisplayName = options.DisplayName,
        Scope = string.Join(" ", options.Scopes),
        ValidTo = this.now().AddDays(ValidToExtensionDays),
        AllOrgs = default,
    };
    pat = await this.client.CreatePatAsync(patTokencreateRequest, cancellationToken).ConfigureAwait(false);
    this.cache.PutPat(options.ToString(), pat);
}
else if (this.ExpiringSoon(pat))
{
    pat = await this.client.RegeneratePatAsync(
        pat,
        this.now().AddDays(ValidToExtensionDays),
        cancellationToken)
    .ConfigureAwait(false);
    this.cache.PutPat(options.ToString(), pat);
}

return pat;
```
How about that?

Reply:
Hello?

Reply:
As discussed offline - I'm good with this. I just didn't like always checking a _new_ Pat if it's close to expiring.